File: stayhome/java/Stayhome.java
import java.util.*;
import java . io .*;

public class Stayhome {

  public static void main (String[] args) {
    int N, M;
    char[][] grid = new char[1000][1000];
    ArrayList<AirState> airInfection = new ArrayList<AirState>();
    CoronaState cor_initial = new CoronaState(0, 0, 0, null);
    SotirisState sot_initial = new SotirisState(0, 0, 0, null);

    //boolean air_inf = false;
    int fx = 0;
    int fy = 0;
    int nl = 0;

    try {
      File input = new File(args[0]);
      BufferedReader reader = new BufferedReader (new FileReader(input));
      String line = null;
      while ((line = reader.readLine()) != null) {
        grid[nl] = line.toCharArray();
        nl++;
      }


    } catch (Exception e) { // Bad practice !
      System.out.println(" Something went wrong : " + e);
    }
    N = nl;
    M = grid[0].length;
    // find starting positions for sotiris, corona and the others
    for (int i = 0; i < N; i++) {
      for (int j = 0; j < M; j++) {
        if (grid[i][j] == 'W') {
          cor_initial = new CoronaState(i, j, 0, null);
        }
        if (grid[i][j] == 'A') {
          airInfection.add(new AirState(i, j, -1));

        }
        if (grid[i][j] == 'S') {
          sot_initial = new SotirisState(i, j, 0, null);

        }
        if (grid[i][j] == 'T') {
          fx = i;
          fy = j;
        }
      }
    }
    // find corona virus spread array
    int [][]covid = coronaBfs(grid, N, M, cor_initial, airInfection);

    // find sotiris final position
    SotirisState kapa = sotirisVsCorona(sot_initial, grid, covid, N, M, fx, fy);
    String path;
    if (kapa == null) {
      System.out.print("IMPOSSIBLE\n");
    }
    else {
      path = findPath(kapa);
      System.out.println(path.length());
      System.out.println(path);
    }



  }
  public static int[][] coronaBfs(char [][]grid, int N, int M, CoronaState virus, ArrayList<AirState> initialStates) {
    int[][] time = new int[N][M];
    Queue<CoronaState> remaining = new ArrayDeque<>();
    Set<CoronaState> seen = new HashSet<>();
    AirState gen_air = new AirState(-1, -1, -1);
    int air_time, cor_time;
    for (int i = 0; i < N; i++) {
      for (int j = 0; j < M; j++) {
        time[i][j] = -1;
      }
    }
    // if airports exist get first one to have access to static vars of airstate class
    if (initialStates.size() != 0)
       gen_air = initialStates.get(0);

    // push to queue
    remaining.add(virus);
    seen.add(virus);
    time[virus.getX()][virus.getY()] = 0;
    while (true) {
      // nowhere to go and  virus has arrived to airport but other airports not infected yet
      if (remaining.isEmpty() && gen_air.getTime() != 0 && gen_air.getInf() != true) {
        // infect all airports
        gen_air.infected();
        for (AirState a : initialStates) {
          int x = a.getX();
          int y = a.getY();
          if (time[x][y] == -1 && (x != virus.getX() || y != virus.getY())) {
            time[x][y] = gen_air.getTime() + 5;
            CoronaState c = new CoronaState(x, y, gen_air.getTime() + 5, virus);
            remaining.add(c);
            seen.add(c);

          }
        }
        continue;
      }
      // nowhere to go, no airports to push -> bye bye...
      else if (remaining.isEmpty()) {
        break;
      }
      virus = remaining.remove();

      air_time = gen_air.getTime();

      cor_time = time[virus.getX()][virus.getY()];

      // virus' first time in airport
      // enable air_time
      if (air_time == 0 && grid[virus.getX()][virus.getY()] == 'A') {
        gen_air.initTime(cor_time);
      }
      // necessary air_time passed, virus now infects all airports
      if (air_time != 0 && cor_time - air_time == 4) {
        gen_air.infected();
        for (AirState a : initialStates) {


          int x = a.getX();
          int y = a.getY();
          if (time[x][y] == -1 && (x != virus.getX() || y != virus.getY())) {
            time[x][y] = air_time + 5;
            CoronaState c = new CoronaState(x, y,  air_time + 5, virus);
            remaining.add(c);
            seen.add(c);
          }
        }
      }
      // find neighbors and infect them
      for (CoronaState n : virus.next()) {
        if (!seen.contains(n) && !n.isBad(grid, N, M)) {
          remaining.add(n);
          seen.add(n);
          time[n.getX()][n.getY()] = time[virus.getX()][virus.getY()] + 2;
        }
      }


    }
    return time;

  }



  private static SotirisState sotirisVsCorona(SotirisState initial, char [][]grid, int[][] covid, int N, int M, int fx, int fy) {

    Queue<SotirisState> remaining = new ArrayDeque<>();
    Set<SotirisState> seen = new HashSet<>();
    remaining.add(initial);
    seen.add(initial);
    while (!remaining.isEmpty()) {
      SotirisState s = remaining.remove();
      int x = s.getX();
      int y = s.getY();
      if (s.isFinal(fx, fy)) return s;

      for (SotirisState n : s.next()) {
        // check if sotiris is not as fast as corona or this grid is a 'X' and other bad situations
        if (!seen.contains(n) && !n.isBad(grid, N, M, covid)) {
          remaining.add(n);
          seen.add(n);
        }
      }

    }
    return null;

  }

  private static String findPath(SotirisState s){
     // go to all previous sotiris states from final pos to start pos
     SotirisState previous = s.getPrevious();
     SotirisState current = s;
     ArrayList<Character> path = new ArrayList<Character>();

     while(previous != null){
       path.add(findMove(previous, current));
       current = previous;
       previous = current.getPrevious();
     }

     StringBuilder builder = new StringBuilder();
     for(int i = path.size() - 1; i >= 0; i--){
       builder.append(path.get(i));
     }
     return builder.toString();
   }

   private static char findMove(SotirisState a, SotirisState b){
     // be careful we want the best lexicographically move
     int x1 = a.getX(), y1 = a.getY();
     int x2 = b.getX(), y2 = b.getY();
     if(y1 == y2){
       if (x2 == x1 + 1){
         return 'D';
       }
       return 'U';
     }
     else {
       if(y2 == y1 + 1){
         return 'R';
       }
       return 'L';
     }
   }


}

File: stayhome/stayhome.py
from collections import deque
import sys



def print_list(list):
    for i in range(len(list)):
        print(list[i], end=" ")
    print("")



def watch_neighbors(pos, grid, N, M):

    neighbors = []
    i = pos[0]
    j = pos[1]
    # down
    if i+1 < N and grid[i+1][j] != "X":
        neighbors.append([i+1, j])

    # left
    if j-1 >= 0 and grid[i][j-1] != "X":
        neighbors.append([i, j-1])
    # right
    if j+1 < M and grid[i][j+1] != "X":
        neighbors.append([i, j+1])

    # up
    if i-1 >= 0 and grid[i-1][j] != "X":
        neighbors.append([i-1, j])

    return neighbors

def coronaBfs(corona, N, M, start_pos, airports, map):
    corQueue = deque(start_pos)
    top = start_pos[0]
    air_time = 0
    cor_time = 0
    air_inf = False
    while (True):
        # nowhere to go and  virus has arrived to airport but other airports not infected yet
        if (len(corQueue) == 0 and air_time != 0 and air_inf != True):
            # infect all airports
            air_inf = True
            for i in airports:
                if (i[0] != top[0] or i[1] != top[1]) and corona[i[0]][i[1]] == -1:
                    corona[i[0]][i[1]] = air_time + 5
                    corQueue.append(i)
            continue
        # nowhere to go, no airports to push -> bye bye...
        elif len(corQueue) == 0:
            break


        top = corQueue.popleft()
        cor_time = corona[top[0]][top[1]]

        # virus' first time in airport
        # enable air_time
        if air_time == 0 and map[top[0]][top[1]] == "A":
            air_time = cor_time

        # necessary air_time passed, virus now infects all airports
        if air_time != 0 and cor_time - air_time == 4:
            air_inf = True
            for i in airports:
                if (i[0] != top[0] or i[1] != top[1]) and corona[i[0]][i[1]] == -1:
                    corona[i[0]][i[1]] = air_time + 5
                    corQueue.append(i)

        # find neighbors and infect them
        neighbors = watch_neighbors(top, corona, N, M)
        #print_list(neighbors)
        for pos in neighbors:
            if corona[pos[0]][pos[1]] == -1:
                corona[pos[0]][pos[1]] = cor_time + 2
                corQueue.append(pos)


    return corona


def sotirisVsCorona(sotiris, corona, N, M, start_pos, end_pos, map):

    sotQueue = deque(start_pos)

    sot_time = 0
    while len(sotQueue) != 0:

        top = sotQueue.popleft()
        sot_time = sotiris[top[0]][top[1]]
        cor_time = corona[top[0]][top[1]]

        # sotiris is not as fast as corona :(
        if cor_time != -1 and cor_time <= sot_time:
            parent[top[0]][top[1]] = [-1, -1]
            continue

        # sotiri, go to neighborhood to meet new people :)
        neighbors = watch_neighbors(top, sotiris, N, M)

        for n in neighbors:
            if sotiris[n[0]][n[1]] == -1:
                sotiris[n[0]][n[1]] = sot_time + 1
                parent[n[0]][n[1]] = top
                sotQueue.append(n)


    return sotiris




def find_next (x1, y1, x2, y2):
    if y1 == y2:
        if x2 == x1 - 1:
            return "U"
        return "D"
    else:
        if y2 == y1 - 1:
            return "L"
        return "R"




def print2Dmap(grid):
    for i in range(N):
        for j in range(M):
            print(grid[i][j], end=" ")
        print("")



def find_path(parent, end_pos):

    path = []
    curr_pos = end_pos
    while(True):
        prev_in_path = parent[curr_pos[0]][curr_pos[1]]
        if prev_in_path == [-1, -1]:
            break

        path.append(find_next(prev_in_path[0], prev_in_path[1], curr_pos[0], curr_pos[1]))
        curr_pos = prev_in_path

    path.reverse()
    return ''.join(path)

if __name__ == "__main__":
    with open(sys.argv[1], "rt") as inputFile:
        map = inputFile.read().split('\n')[:-1]
    N = len(map)
    M = len(map[0])
    airports = []
    cor_start_pos = []
    sot_start_pos = []

    corona  = [[-1 for j in range(M)] for i in range(N)]
    sotiris = [[-1 for j in range(M)] for i in range(N)]
    parent = [[[-1,-1] for j in range(M)] for i in range(N)]

    #print2Dmap(map)


    #find all necessary info
    for i in range(N):
        for j in range(M):
            if map[i][j] == "S":
                sot_start_pos.append((i,j))
                sotiris[i][j] = 0
            elif map[i][j] == "T":
                sot_end_pos = [i, j]
            elif map[i][j] == "A":
                airports.append((i,j))
            elif map[i][j] == "W":
                cor_start_pos.append((i,j))
                corona[i][j] = 0
            elif map[i][j] == "X":
                corona[i][j] = "X"
                sotiris[i][j] = "X"
    #print(cor_start_pos[0])
    #print_list(airports)
    corona = coronaBfs(corona, N, M, cor_start_pos, airports, map)
    print2Dmap(corona)
    sotiris = sotirisVsCorona(sotiris, corona, N, M, sot_start_pos, sot_end_pos, map)
    #print2Dmap(sotiris)



    path = find_path(parent, sot_end_pos)
    if (len(path) == 0):
        print("IMPOSSIBLE")
    else:
        print(len(path))
        print(path)
